driver-sync/src/test/functional/com/mongodb/internal/connection/OidcAuthenticationProseTests.java

```diff
@@ -256,118 +253,210 @@ public void test3p1AuthFailsWithCachedToken() throws ExecutionException, Interru
             failCommand(391, 1, "find");
             // #. Perform a find operation that succeeds.
             performFind(mongoClient);
         }
-        assertEquals(2, onRequest.invocations.get());
+        assertEquals(2, callback.invocations.get());
+    }
+
+    @Test
+    public void test4p2ReadCommandsFailIfReauthenticationFails() {
+        // Create a `MongoClient` whose OIDC callback returns one good token
+        // and then bad tokens after the first call.
+        TestCallback wrappedCallback = createCallback();
+        OidcCallback callback = (context) -> {
+            OidcCallbackResult result1 = wrappedCallback.callback(context);
+            return new OidcCallbackResult(wrappedCallback.getInvocations() > 1 ? "bad" : result1.getAccessToken());
+        };
+        MongoClientSettings clientSettings = createSettings(callback);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            performFind(mongoClient);
+            failCommand(391, 1, "find");
+            assertCause(MongoCommandException.class,
+                    "Command failed with error 18",
+                    () -> performFind(mongoClient));
+        }
+        assertEquals(2, wrappedCallback.invocations.get());
+    }
+
+    @Test
+    public void test4p3WriteCommandsFailIfReauthenticationFails() {
+        // Create a `MongoClient` whose OIDC callback returns one good token
+        // and then bad tokens after the first call.
+        TestCallback wrappedCallback = createCallback();
+        OidcCallback callback = (context) -> {
+            OidcCallbackResult result1 = wrappedCallback.callback(context);
+            return new OidcCallbackResult(
+                    wrappedCallback.getInvocations() > 1 ? "bad" : result1.getAccessToken());
+        };
+        MongoClientSettings clientSettings = createSettings(callback);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            performInsert(mongoClient);
+            failCommand(391, 1, "insert");
+            assertCause(MongoCommandException.class,
+                    "Command failed with error 18",
+                    () -> performInsert(mongoClient));
+        }
+        assertEquals(2, wrappedCallback.invocations.get());
+    }
+
+    private static void performInsert(final MongoClient mongoClient) {
+        mongoClient
+                .getDatabase("test")
+                .getCollection("test")
+                .insertOne(Document.parse("{ x: 1 }"));
+    }
+
+    @Test
+    public void test5p1Azure() {
+        assumeTrue(getOidcEnv().equals("azure"));
+        String oidcUri = getOidcUri();
+        assertFalse(oidcUri.contains("@"));
+        MongoClientSettings clientSettings = createSettings(oidcUri, createCallback(), null);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            // #. Perform a find operation that succeeds.
+            performFind(mongoClient);
+        }
+    }
```

Comment:
I assume this test corresponds to https://github.com/mongodb/specifications/blob/master/source/auth/tests/mongodb-oidc.md#51-azure-with-no-username. It requires a valid `TOKEN_RESOURCE` to be set, but the test neither sets it explicitly, nor explicitly verifies that `TOKEN_RESOURCE` is set. Instead, the test relies on `getOidcUri()` having it. Such an approach means that if the assumption that `getOidcUri()` has `TOKEN_RESOURCE` is violated, the test will fail with an exception coming from the driver code (as opposed to coming from the test code), which will be confusing.

The approach taken in the PR for checking that `getOidcUri()` does not have a user name specified, is different - there is an explicit check `assertFalse(oidcUri.contains("@"))` for that. Let's do the same for `TOKEN_RESOURCE`, only checking it based on `clientSettings` is clearer and less error-prone than checking it by looking at the `oidcUri` string.

P.S. If we have other tests with similar situations (`test5p1AzureFails` is another one, maybe there are more), we should also introduce the checks there.

Reply:
I have added checks, but I do not think they are strictly needed. We generally trust that that the test runner and environment is correctly set up, so for example, an "assumption" would just disable the test. However here, I think it is fine to have these checks, because these two Azure tests are somewhat different from the rest, just to make it clearer to the reader that no mistakes were made in setup.